File: qlat/pylib/qlat/fermion_action.py
import qlat.cqlat as c

class FermionAction:

    def __init__(self, *, mass, ls, m5, mobius_scale = 1.0, omega = None):
        assert isinstance(mass, float)
        assert isinstance(ls, int)
        assert isinstance(m5, float)
        if omega is None:
            self.cdata = c.mk_fermion_action_mobius(mass, ls, m5, mobius_scale)
        else:
            assert isinstance(omega, list)
            assert ls == len(omega)
            self.cdata = c.mk_fermion_action_zmobius(mass, m5, omega)

    def __del__(self):
        assert isinstance(self.cdata, int)
        c.free_fermion_action(self)

    def __imatmul__(self, v1):
        assert isinstance(v1, FermionAction)
        c.set_fermion_action(self, v1)
        return self

    def mass(self):
        return c.get_mass_fermion_action(self)

    def ls(self):
        return c.get_ls_fermion_action(self)

    def m5(self):
        return c.get_m5_fermion_action(self)

    def omega(self):
        return c.get_omega_fermion_action(self)

    def mobius_scale(self):
        return c.get_mobius_scale_fermion_action(self)

File: qlat/cqlat/fermion-action.cpp
#include "lib.h"

EXPORT(free_fermion_action, {
  using namespace qlat;
  return free_obj<FermionAction>(args);
})

EXPORT(set_fermion_action, {
  using namespace qlat;
  return set_obj<FermionAction>(args);
})

EXPORT(get_ls_fermion_action, {
  using namespace qlat;
  PyObject* p_fa = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_fa)) {
    return NULL;
  }
  const FermionAction& fa = py_convert_type<FermionAction>(p_fa);
  return py_convert(fa.ls);
})

EXPORT(get_omega_fermion_action, {
  using namespace qlat;
  PyObject* p_fa = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_fa)) {
    return NULL;
  }
  const FermionAction& fa = py_convert_type<FermionAction>(p_fa);
  if (fa.is_using_zmobius) {
    std::vector<ComplexD> omega(fa.bs.size());
    for (Int i = 0; i < (int)omega.size(); ++i) {
      omega[i] = 1.0 / (fa.bs[i] + fa.cs[i]);
    }
    return py_convert(omega);
  } else {
    Py_RETURN_NONE;
  }
})

EXPORT(get_mobius_scale_fermion_action, {
  using namespace qlat;
  PyObject* p_fa = NULL;
  if (!PyArg_ParseTuple(args, "O", &p_fa)) {
    return NULL;
  }
  const FermionAction& fa = py_convert_type<FermionAction>(p_fa);
  if (fa.is_using_zmobius) {
    qassert(fa.mobius_scale == 0.0);
  } else {
    qassert(fa.mobius_scale != 0.0);
  }
  return py_convert(fa.mobius_scale);
})
